### DISCUSSION TITLE
airflow.exceptions.AirflowException: missing keyword argument 'slack_conn_id'

### DISCUSSION THREAD

**Discussion by M-Waleed-Khan** [apache/airflow#37535] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.7.2

### What happened?

I was using Apache 2.2.4 and everything was working fine. I recently upgraded my env to Apache 2.7.2. Now, In one of my dag failure I am not able to get the slack notification because it says the above error.

 I am using this code below
 
 ```
failed_alert = SlackAPIPostOperator(
        task_id='slack_failed',
        channel=channel,
        token=token,
        username=user,
        text="""
                :red_circle: Task Failed. 
                *Task*: {task}  
                *Dag*: {dag} 
                *Execution Time*: {exec_date}  
                *Error* : {errormessage}
                *Log Url*: {log_url} 
        """.format(
            task=context.get('task_instance').task_id,
            dag=context.get('task_instance').dag_id,
            ti=context.get('task_instance'),
            exec_date=context.get('execution_date'),
            log_url=context.get('task_instance').log_url,
            errormessage=context.get('exception')
        )
    )

    return failed_alert.execute(context=context)`
```

The error that I am getting is this

```
Traceback (most recent call last):
  File "/usr/local/airflow/.local/lib/python3.11/site-packages/airflow/models/taskinstance.py", line 1702, in _run_finished_callback
    callback(context)
  File "/usr/local/airflow/dags/scripts/slack_logger.py", line 11, in slack_failed_task
    failed_alert = SlackAPIPostOperator(
                   ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/airflow/.local/lib/python3.11/site-packages/airflow/models/baseoperator.py", line 437, in apply_defaults
    result = func(self, **kwargs, default_args=default_args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/airflow/.local/lib/python3.11/site-packages/airflow/providers/slack/operators/slack.py", line 131, in __init__
    super().__init__(method=self.method, **kwargs)
  File "/usr/local/airflow/.local/lib/python3.11/site-packages/airflow/models/baseoperator.py", line 420, in apply_defaults
    raise AirflowException(f"missing keyword argument {missing_args.pop()!r}")
airflow.exceptions.AirflowException: missing keyword argument 'slack_conn_id'
```

### How to reproduce

be calling the SlackAPIPostOperator

### Operating System

Mac and AWS MWAA both same

### Versions of Apache Airflow Providers

Current version
apache-airflow-providers-slack==8.1.0
Previous working version
apache-airflow-providers-slack==4.2.0

### Deployment

Amazon (AWS) MWAA

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by Taragolis** [comment]
> I was using Apache 2.2.4 and everything was working fine. I recently upgraded my env to Apache 2.7.2.

I guess you also upgrade the Slack Provider. Because in v6 of provider `token` was deprecated, and in v8 `slack_conn_id` become mandatory, see [Slack Provider Changelog](https://airflow.apache.org/docs/apache-airflow-providers-slack/stable/changelog.html#id17)

> I am using this code below

Please don't use this code, and avoid (ever) manually invoke `execute` method of the provider, it is only should be executed by the Airflow Worker, behind the scene a lot of stuff happen and in some cases some operators might work but in another it breaks with some random error.

I've just wondering, what is original source of that code because time to time we saw that kind of usage of the operator, e.g. execute one operator in another, or use in callbacks. And really want to prevent this kind of usage but don't know how 😞 

But there is a bright side, since you use Airflow 2.6+ you also could use Notifiers, it is improvement version of callbacks, and Slack Provider have 2 of them. [airflow.providers.slack.notifications.slack.SlackNotifier](https://airflow.apache.org/docs/apache-airflow-providers-slack/stable/notifications/slack_notifier_howto_guide.html) should be a builtin replacement to your callback.

If you want to customise callback, you always could create your own Notifier, as a reference you could use source code of [SlackNotifier](https://airflow.apache.org/docs/apache-airflow-providers-slack/stable/_modules/airflow/providers/slack/notifications/slack.html#SlackNotifier)

**Reply by M-Waleed-Khan** [reply]
HI @Taragolis Thanks for your respone

as I mentioned above these are the slack providers :
Current version
apache-airflow-providers-slack==8.1.0
Previous working version
apache-airflow-providers-slack==4.2.0

**Reply by M-Waleed-Khan** [reply]
and since I am using 8.1.0 I read the documentation it says the slack_conn_id is optional ?

![image](https://github.com/apache/airflow/assets/48055930/9c3b0fb7-85dc-440a-b3c3-702d0ea12f81)

**Reply by M-Waleed-Khan** [reply]
and Why I am using the code this way is because this is how my code is structured

on_failure_callback=slack_logger.slack_failed_task,

now slack_logger is a file which contains a generic code for all failures handlers of my Dags

```
`def slack_failed_task(context):
    secret = etl.get_secrets("secret_name")
    secret = json.loads(secret)
    user, channel, token = [secret.get(key) for key in secret.keys()]
    failed_alert = SlackAPIPostOperator(
        task_id='slack_failed',
        channel=channel,
        token=token,
        username=user,
        text="""
                :red_circle: Task Failed. 
                *Task*: {task}  
                *Dag*: {dag} 
                *Execution Time*: {exec_date}  
                *Error* : {errormessage}
                *Log Url*: {log_url} 
        """.format(
            task=context.get('task_instance').task_id,
            dag=context.get('task_instance').dag_id,
            ti=context.get('task_instance'),
            exec_date=context.get('execution_date'),
            log_url=context.get('task_instance').log_url,
            errormessage=context.get('exception')
        )
    )

    return failed_alert.execute(context=context)`
```

**Reply by Taragolis** [reply]
> and since I am using 8.1.0 I read the documentation it says the slack_conn_id is optional ?

Simply answer: it was a bug into the documentation, which I guess already fixed into the new version.
If you need to use token directly, you have to downgrade to `apache-airflow-providers-slack==7.3.2`.

> and Why I am using the code this way is because this is how my code is structured

The point here that you should avoid manually execute operators, operators executed by the airflow worker. Better consider to use Hooks that it their main purposes, you could use it whatever you wanted: Into the custom operators, into the task flow operators, into Notifiers/Callbacks.

**Reply by potiuk** [reply]
Well. just to reiterate what @Taragolis wrote. You are using operators wrongly. They are not supposed to be used this way and what worked before was accidentally working, but it was not intention for operator to be used this way, so the only good way for you is to adjust 
and do it properly. If you want to communicate with Slack in more stable way from your code - use SlackHook. Hooks are implementing "airflow way" of communicating with external services, and they are supposed to be used for that (this is how Operators use them). Operators are there to "embed" communication with external systems inside the "task" that is supposed to be executed exclusively inside the DAG - with a LOT of setup that is done when the DAG is parsed and even more, when particular task is chosen to be executed.

There is no intention and if you look at airflow concepts and documentation for Operators to be used in other ways. If you thought it's not teh case, you made wrong assesment, and the best way for you to fix it is to come back to the way Hooks, and Operators were intended to use, rather than insist that "your way" should be supported by Airflow even if it's against the original intentions of the whole platform. 


Surely you can use screwdriver to put the nail into a wooden stick if you hit hard and long enough, but you can break your finger and destroy few things along the way. It's simply  better to use hammer for that.

**Reply by M-Waleed-Khan** [reply]
since I am new to airflow slack notification can you please give me a small example code thing or direction how can I change the way I am using the operator now and how it should be done to achieve what I am achieving right now
Thanks!!

**Reply by potiuk** [reply]
Just instantiate Hook and use it. Same way you used Operator - find the docstring. And look at the source code of Operator to see how operators use it. Airflow is open-source and you can always look at the code, it's easily searchable and browsable with Github. It is a poor usage of maintainer's time if they would mostly look for the code and copy-pasted it for new users. We have tens of thousands of users and if we do just that, we would not have time to do anything else.

**Comment by rud0812** [comment]
Hi @M-Waleed-Khan! I was having this issue today and managed to solved it. I was using airflow 2.5.2 with no issues and got the same error when updated to airflow 2.7.3. 

Just some background. When using airflow 2.5.2, I followed this [tutorial](https://medium.com/datareply/integrating-slack-alerts-in-airflow-c9dcd155105) on how to set the slack connection. Basically I used the following function and in my dags default_args just added the line `'on_failure_callback': task_fail_slack_alert`:

```python
def task_fail_slack_alert(context):
    SLACK_CONN_ID = 'slack'
    slack_webhook_token = BaseHook.get_connection(SLACK_CONN_ID).password
    slack_msg = """
            :red_circle: Task Failed. 
            *Task*: {task}  
            *Dag*: {dag} 
            *Execution Time*: {exec_date}  
            *Log Url*: {log_url} 
            """.format(
            task=context.get('task_instance').task_id,
            dag=context.get('task_instance').dag_id,
            ti=context.get('task_instance'),
            exec_date=context.get('execution_date'),
            log_url=context.get('task_instance').log_url,
        )
    failed_alert = SlackWebhookOperator(
        task_id='slack_test',
        http_conn_id='slack',
        webhook_token=slack_webhook_token,
        message=slack_msg,
        username='airflow')
    return failed_alert.execute(context=context)
``` 

Seems like after updating to 2.7.3, it uses apache-airflow-providers-slack==8.1.0 where `http_conn_id` got deprecated and now you need to use `slack_webhook_conn_id`. So basically you can swap it right out of the box and also delete the webhook_token line which doesn't do anything anymore. New function looks as following and works:
```python
def task_fail_slack_alert(context):
    SLACK_CONN_ID = 'slack'
    slack_webhook_token = BaseHook.get_connection(SLACK_CONN_ID).password
    slack_msg = """
            :red_circle: Task Failed. 
            *Task*: {task}  
            *Dag*: {dag} 
            *Execution Time*: {exec_date}  
            *Log Url*: {log_url} 
            """.format(
            task=context.get('task_instance').task_id,
            dag=context.get('task_instance').dag_id,
            ti=context.get('task_instance'),
            exec_date=context.get('execution_date'),
            log_url=context.get('task_instance').log_url,
        )
    failed_alert = SlackWebhookOperator(
        task_id='slack_test',
        slack_webhook_conn_id=SLACK_CONN_ID,
        message=slack_msg,
        username='airflow')
    return failed_alert.execute(context=context)
```

**Reply by Taragolis** [reply]
@kaxil wrote this tutorial more than 5 years ago, many things was changed from that time. One major changes is Notifiers which added into the Airflow 2.6

**Reply by kaxil** [reply]
Yup I would suggest using Slack notifier, I might try to update my tutorial at some point too if I manage to find some time :)
